File: iosApp/iosApp/Media/MediaPlaybackCoordinator.swift
import Foundation
import Combine
import VocableShared

/// Arms delayed phrase media playback and drives fullscreen overlay state.
final class MediaPlaybackCoordinator: ObservableObject {
    enum Phase: Equatable {
        case idle
        case armed
        case playing
    }

    @Published private(set) var phase: Phase = .idle
    @Published private(set) var activePhrase: PhraseDisplayModel?
    @Published var isPaused: Bool = false
    @Published var showCenterControls: Bool = false
    @Published var showExitControl: Bool = false

    private var armedPhrase: PhraseDisplayModel?
    private var idleWorkItem: DispatchWorkItem?
    private let settings = AppSettings.shared

    func onPhraseSelected(_ phrase: PhraseDisplayModel) {
        idleWorkItem?.cancel()
        idleWorkItem = nil

        guard phrase.style?.triggersDelayedPlayback() == true else {
            if phase == .armed {
                phase = .idle
                armedPhrase = nil
            }
            return
        }

        if phase == .playing {
            return
        }

        if armedPhrase?.id == phrase.id, phase == .armed {
            return
        }

        armedPhrase = phrase
        phase = .armed

        let work = DispatchWorkItem { [weak self] in
            self?.beginPlayback()
        }
        idleWorkItem = work
        DispatchQueue.main.asyncAfter(
            deadline: .now() + settings.mediaPlaybackDelay,
            execute: work
        )
    }

    func cancelPending() {
        idleWorkItem?.cancel()
        idleWorkItem = nil
        if phase == .armed {
            phase = .idle
            armedPhrase = nil
        }
    }

    func beginPlayback() {
        guard let phrase = armedPhrase,
              phrase.style?.triggersDelayedPlayback() == true else {
            phase = .idle
            return
        }
        idleWorkItem = nil
        armedPhrase = nil
        activePhrase = phrase
        isPaused = false
        showCenterControls = false
        showExitControl = false
        phase = .playing
        TTSManager.shared.stop()
    }

    func togglePlayPause() {
        isPaused.toggle()
    }

    func stopEarly() {
        tearDown()
    }

    func onNaturalEnd() {
        tearDown()
    }

    private func tearDown() {
        idleWorkItem?.cancel()
        idleWorkItem = nil
        armedPhrase = nil
        activePhrase = nil
        isPaused = false
        showCenterControls = false
        showExitControl = false
        phase = .idle
    }
}

File: web/src/media/mediaPlaybackCoordinator.ts
import type { PhraseDisplay } from '../data/types'
import { hasPhraseMedia } from '../data/phraseStyle'

export type MediaPlaybackPhase = 'idle' | 'armed' | 'playing'

export type MediaPlaybackListener = (state: MediaPlaybackState) => void

export interface MediaPlaybackState {
  phase: MediaPlaybackPhase
  activePhrase: PhraseDisplay | null
  isPaused: boolean
  showCenterControls: boolean
  showExitControl: boolean
}

const initialState: MediaPlaybackState = {
  phase: 'idle',
  activePhrase: null,
  isPaused: false,
  showCenterControls: false,
  showExitControl: false,
}

export class MediaPlaybackCoordinator {
  private state: MediaPlaybackState = { ...initialState }
  private listeners = new Set<MediaPlaybackListener>()
  private idleTimer: ReturnType<typeof setTimeout> | null = null
  private armedPhrase: PhraseDisplay | null = null
  private getDelay: () => number

  constructor(getDelay: () => number) {
    this.getDelay = getDelay
  }

  subscribe(listener: MediaPlaybackListener): () => void {
    this.listeners.add(listener)
    listener(this.state)
    return () => this.listeners.delete(listener)
  }

  private emit() {
    const snapshot = { ...this.state }
    this.listeners.forEach((l) => l(snapshot))
  }

  onPhraseSelected(phrase: PhraseDisplay) {
    if (this.idleTimer) {
      clearTimeout(this.idleTimer)
      this.idleTimer = null
    }

    if (!hasPhraseMedia(phrase.style)) {
      if (this.state.phase === 'armed') {
        this.armedPhrase = null
        this.state = { ...initialState }
        this.emit()
      }
      return
    }

    if (this.state.phase === 'playing') return

    if (this.state.phase === 'armed' && this.armedPhrase?.id === phrase.id) {
      return
    }

    this.armedPhrase = phrase
    this.state = { ...this.state, phase: 'armed' }
    this.emit()

    this.idleTimer = setTimeout(() => {
      this.beginPlayback()
    }, this.getDelay() * 1000)
  }

  cancelPending() {
    if (this.idleTimer) {
      clearTimeout(this.idleTimer)
      this.idleTimer = null
    }
    if (this.state.phase === 'armed') {
      this.armedPhrase = null
      this.state = { ...initialState }
      this.emit()
    }
  }

  beginPlayback() {
    const phrase = this.armedPhrase
    if (!phrase || !hasPhraseMedia(phrase.style)) {
      this.state = { ...initialState }
      this.emit()
      return
    }
    this.idleTimer = null
    this.armedPhrase = null
    this.state = {
      phase: 'playing',
      activePhrase: phrase,
      isPaused: false,
      showCenterControls: false,
      showExitControl: false,
    }
    this.emit()
  }

  togglePlayPause() {
    if (this.state.phase !== 'playing') return
    this.state = { ...this.state, isPaused: !this.state.isPaused }
    this.emit()
  }

  setShowCenterControls(show: boolean) {
    if (this.state.showCenterControls === show) return
    this.state = { ...this.state, showCenterControls: show }
    this.emit()
  }

  setShowExitControl(show: boolean) {
    if (this.state.showExitControl === show) return
    this.state = { ...this.state, showExitControl: show }
    this.emit()
  }

  stopEarly() {
    this.tearDown()
  }

  onNaturalEnd() {
    this.tearDown()
  }

  private tearDown() {
    if (this.idleTimer) {
      clearTimeout(this.idleTimer)
      this.idleTimer = null
    }
    this.armedPhrase = null
    this.state = { ...initialState }
    this.emit()
  }
}
